### DISCUSSION TITLE
Full QualifiedName / CanonicalName from typeMirror

### DISCUSSION THREAD

**Discussion by NilsBaumgartner1994** [pmd/pmd#4572] [Q&A]
A new question arises from: https://github.com/pmd/pmd/discussions/4571
I want to get the full qualified name / canonical name of a field/parameter of which i may use JTypeMirror.

A problem arises when i try to get the canonical name for generics. I think that i need to get the canonical name for each type Argument but maybe this isn't possible to get from JTypeMirror alone?

Field to analyse:
```java
private java.util.Hashtable<javax.swing.JSlider, CodePiece> hashMap;
```

Here is the current implementation of getting the full qualified name.
```java
private String getQualifiedNameFromTypeMirror(JTypeMirror typeMirror){
        // Invoke the method on the interfaceType object

        JTypeDeclSymbol symbol = typeMirror.getSymbol();

        if (symbol instanceof JClassSymbol) {
            String fullQualifiedName = ((JClassSymbol) symbol).getCanonicalName();
            return fullQualifiedName;
        }
        return null;
    }
```

**Comment by oowekyala** [comment]
You can downcast the `JTypeMirror` to a `JClassType` instead of operating on the symbol. Then call `JClassType#getTypeArgs()` to get the type arguments.

**Reply by NilsBaumgartner1994** [reply]
Yes this is one possible way. Its a bit of a workaround and then i need to call for each typeArgs the same canonicalName. But this could solve the problem.

**Reply by oowekyala** [reply]
I wouldn't term this a workaround. The type API reflects everything about types.  Looking at your code in #4571, you would have an easier time querying the type directly using `JClassType#getSuperClass`, `getSuperInterfaces`, and `getDeclaredFields` instead of looking through the AST nodes. From an `ASTClassOrInterfaceDeclaration`, you can use [`getTypeMirror`](https://docs.pmd-code.org/apidocs/pmd-java/7.0.0-SNAPSHOT/net/sourceforge/pmd/lang/java/ast/ASTAnyTypeDeclaration.html#getTypeMirror()) to get its mirror and query its fields and such. You won't even need the downcast to JClassType.

**Reply by oowekyala** [reply]
Alternatively if you just want a string from a type, use [TypePrettyPrint](https://docs.pmd-code.org/apidocs/pmd-java/7.0.0-SNAPSHOT/net/sourceforge/pmd/lang/java/types/TypePrettyPrint.html). 

Also the asterisk you're removing in https://github.com/pmd/pmd/discussions/4571#discussioncomment-5973058 means that your type is unresolved, ie some referenced classes are not on the auxclasspath. You should configure the auxclasspath correctly to avoid having to do that.

**Reply by NilsBaumgartner1994** [reply]
> I wouldn't term this a workaround. The type API reflects everything about types. Looking at your code in https://github.com/pmd/pmd/discussions/4571, you would have an easier time querying the type directly using JClassType#getSuperClass, getSuperInterfaces, and getDeclaredFields instead of looking through the AST nodes. From an ASTClassOrInterfaceDeclaration, you can use [getTypeMirror](https://docs.pmd-code.org/apidocs/pmd-java/7.0.0-SNAPSHOT/net/sourceforge/pmd/lang/java/ast/ASTAnyTypeDeclaration.html#getTypeMirror()) to get its mirror and query its fields and such. You won't even need the downcast to JClassType.

Would this only suite for Classes but i also want Interfaces to analyse?

> Alternatively if you just want a string from a type, use [TypePrettyPrint](https://docs.pmd-code.org/apidocs/pmd-java/7.0.0-SNAPSHOT/net/sourceforge/pmd/lang/java/types/TypePrettyPrint.html).
> 
> Also the asterisk you're removing in [#4571 (comment)](https://github.com/pmd/pmd/discussions/4571#discussioncomment-5973058) means that your type is unresolved, ie some referenced classes are not on the auxclasspath. You should configure the auxclasspath correctly to avoid having to do that.

Nice thing about TypePrettyPrint :-)
Yes that is correct, that these are not in the auxclasspath. My goal is to identify Data-Clumps automatically. The problem is pulling a repo from a random project gives you not always the information needed about the auxclasspath. Therefore my approach focuses only on the source code.
 
One thing i dont understand is why for example this Type has still an asterik, since the import is explicit given.

```java
import org.flywaydb.core.api.configuration.ClassicConfiguration;
public class Flyway {
    private ClassicConfiguration withoutExplicitType;
}
```

**Reply by oowekyala** [reply]
> Would this only suite for Classes but i also want Interfaces to analyse?

This would handle any type of class, including interfaces, enums, records and annotations. There are methods on `JClassType` like `isEnum`, `isInterface`.

> Yes that is correct, that these are not in the auxclasspath. My goal is to identify Data-Clumps automatically. The problem is pulling a repo from a random project gives you not always the information needed about the auxclasspath. Therefore my approach focuses only on the source code.

Maybe the repo uses a build tool like maven/gradle and you can run the build before running PMD? That's what we do in our CI. Currently PMD cannot see outside of a given source file without using the auxclasspath, so without this data, the best you can do is guess, and your guess might well be wrong. For instance in https://github.com/pmd/pmd/discussions/4571#discussioncomment-5973058 you assume it's in the current package, but maybe no such class exists in the current package. The class may be inherited from a supertype or something. The only way to know for sure is to populate the auxclasspath somehow.

> One thing i dont understand is why for example this Type has still an asterik, since the import is explicit given.

The asterisk does not mean that we don't know the full name of the type. It means we don't know anything more about the type than its name, so the symbol will return inaccurate data (eg we don't know the fields of `ClassicConfiguration` or its methods. We're not even sure of its package name). Maybe we could have an option to remove the asterisk in TypePrettyPrinter... but the better solution is really to avoid unresolved types.
